### PR TITLE
Adding support for whitelisting of user-defined managed identities

### PR DESCRIPTION
**Pull request by dmeytin**
**Reason for Change**:
In some cases a user-defined managed identity could be removed from the AAD

**Issue Fixed**:
This PR is adding whitelist of user defined managed identities what will never be removed from AAD

**Notes for Reviewers**:
The change is as following:
- mic main  accepts new cmd parameter *whiteListedIdentitiesParam* - comma separated list of identities that shouldn't be deleted from AAD
- at *updateUserMSI* function we compare *removeUserAssignedMSIIDs* and *whiteListedIdentities* and only missing values are removed from AAD
- helm chart and Readme are changed accordingly

This feature is extremely important for migrating existing production clusters to aad-pod-identity.
In particular, all our existing deployments will use user-defined identity as one per team.
Reuse of the same identity to multiple pods caused the identity to be deleted from AAD when it still in use. 
This change will solve the issue.

Best regards,
Dmitry

### CODE REVIEW

File: pkg/mic/mic.go [hunk 1]
```diff
@@ -794,7 +801,15 @@ func (c *Client) updateUserMSI(newAssignedIDs map[string]aadpodid.AzureAssignedI
 	addUserAssignedMSIIDs := c.getUniqueIDs(nodeTrackList.addUserAssignedMSIIDs)
 	removeUserAssignedMSIIDs := c.getUniqueIDs(nodeTrackList.removeUserAssignedMSIIDs)
 
-	err := c.CloudClient.UpdateUserMSI(addUserAssignedMSIIDs, removeUserAssignedMSIIDs, nodeOrVMSSName, nodeTrackList.isvmss)
+	// The whitelisted identities (user-defined managed identities) shouldn't be deleted from AAD, but only from Kubernetes
+	approvedRemoveUserAssignedMSIIDs := make([]string, 0)
+	for _, item := range removeUserAssignedMSIIDs {
+		if _, ok := c.WhiteListedIdentities[item]; !ok {
```
Comment (idanlevin) [inline pkg/mic/mic.go]: are IDs case-sensitive? if not, we should probably check it case-insensitive

File: pkg/mic/mic_test.go [hunk 2]
```diff
@@ -601,18 +601,20 @@ func NewMICTestClient(eventCh chan aadpodid.EventType,
 	podClient *TestPodClient,
 	nodeClient *TestNodeClient,
 	eventRecorder *TestEventRecorder, isNamespaced bool,
-	createDeleteBatch int64) *TestMICClient {
+	createDeleteBatch int64,
+	whiteListedIdentities map[string]bool) *TestMICClient {
```
Comment (idanlevin) [inline pkg/mic/mic_test.go]: we should probably add a dedicated test the makes sure whitelisted identities are not deleted.

File: charts/aad-pod-identity/README.md [hunk 3]
```diff
@@ -162,6 +162,7 @@ The following tables list the configurable parameters of the aad-pod-identity ch
...
 | `mic.probePort`                          | Override http liveliness probe port                                                                                                                                                                              | If not provided, default port is `8080`                  |
 | `mic.syncRetryDuration`                  | Override interval in seconds at which sync loop should periodically check for errors and reconcile                                                                                                               | If not provided, default value is `3600s`                |
+| `mic.whiteListedIdentitiesParam`         | List of whitelisted user-defined identities that shouldn't be deleted from AAD. Note: the list should be comma separated                                                                                         | If not provided, default value is empty           |
```
Comment (idanlevin) [inline charts/aad-pod-identity/README.md]: `whitelist` might be an inaccurate terminology to use, this case isn't about safe/allow vs unsafe/block identities, we basically want to exclude some identities from being deleted. I would go with something more explicit like `mic.excludeFromDeletionIdentitiesParam` or similar.

File: pkg/mic/mic.go [hunk 4]
```diff
@@ -794,7 +801,15 @@ func (c *Client) updateUserMSI(newAssignedIDs map[string]aadpodid.AzureAssignedI
...
+			approvedRemoveUserAssignedMSIIDs = append(approvedRemoveUserAssignedMSIIDs, item)
```
Comment (kkmsft) [inline pkg/mic/mic.go]: Could we add an e2e for this, please.

File: charts/aad-pod-identity/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkmsft) [reply charts/aad-pod-identity/README.md]: Please also specify the format in which the parameters should be specified.

File: cmd/mic/main.go [hunk 5]
```diff
@@ -62,6 +64,9 @@ func main() {
 	// Client QPS is used to configure the client-go QPS throttling and bursting.
 	flag.Float64Var(&clientQPS, "clientQps", 5, "Client QPS used for throttling of calls to kube-api server")
 
+	//Identities that should be never removed from Azure AD (used defined managed identities)
+	flag.StringVar(&whiteListedIdentitiesParam, "whiteListedIdentitiesParam", "", "whitelisted identities to prevent deletion from Azure AD")
```
[the remark this answers is not in the document]
Comment (dmeytin) [reply cmd/mic/main.go]: Done

File: go.sum [hunk 6]
```diff
@@ -199,6 +199,7 @@ golang.org/x/sync v0.0.0-20181108010431-42b317875d0f/go.mod h1:RxMgew5VJxzue5/jJ
 golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
 golang.org/x/sync v0.0.0-20190227155943-e225da77a7e6 h1:bjcUS9ztw9kFmmIxJInhon/0Is3p+EHBKNgquIzo1OI=
 golang.org/x/sync v0.0.0-20190227155943-e225da77a7e6/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
+golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e h1:vcxGaoTs7kV8m5Np9uUNQin4BrLOthgV7252N8V+FwY=
```
[the remark this answers is not in the document]
Comment (dmeytin) [reply go.sum]: Checking in go.sum file can help others to verify that they are using the same modules as you, me or any other developer.
It isn't required by considered to be a good practice.
If you prefer, I'll remove it...

File: pkg/mic/mic.go [hunk 7]
```diff
@@ -794,7 +801,15 @@ func (c *Client) updateUserMSI(newAssignedIDs map[string]aadpodid.AzureAssignedI
 	addUserAssignedMSIIDs := c.getUniqueIDs(nodeTrackList.addUserAssignedMSIIDs)
 	removeUserAssignedMSIIDs := c.getUniqueIDs(nodeTrackList.removeUserAssignedMSIIDs)
 
-	err := c.CloudClient.UpdateUserMSI(addUserAssignedMSIIDs, removeUserAssignedMSIIDs, nodeOrVMSSName, nodeTrackList.isvmss)
+	// The whitelisted identities (user-defined managed identities) shouldn't be deleted from AAD, but only from Kubernetes
```
[the remark this answers is not in the document]
Comment (dmeytin) [reply pkg/mic/mic.go]: Done

File: charts/aad-pod-identity/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmeytin) [reply charts/aad-pod-identity/README.md]: Done

[the remark this answers is not in the document]
Comment (dmeytin) [reply charts/aad-pod-identity/README.md]: Done

File: pkg/mic/mic.go [hunk 8]
```diff
@@ -117,17 +118,23 @@ func NewMICClient(cloudconfig string, config *rest.Config, isNamespaced bool, sy
 	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: clientSet.CoreV1().Events("")})
 	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: aadpodid.CRDGroup})
 
+	whiteListedIdentitiesMap := make(map[string]bool)
```
[the remark this answers is not in the document]
Comment (dmeytin) [reply pkg/mic/mic.go]: Done

File: pkg/mic/mic.go [hunk 9]
```diff
@@ -794,7 +801,15 @@ func (c *Client) updateUserMSI(newAssignedIDs map[string]aadpodid.AzureAssignedI
 	addUserAssignedMSIIDs := c.getUniqueIDs(nodeTrackList.addUserAssignedMSIIDs)
 	removeUserAssignedMSIIDs := c.getUniqueIDs(nodeTrackList.removeUserAssignedMSIIDs)
 
-	err := c.CloudClient.UpdateUserMSI(addUserAssignedMSIIDs, removeUserAssignedMSIIDs, nodeOrVMSSName, nodeTrackList.isvmss)
+	// The whitelisted identities (user-defined managed identities) shouldn't be deleted from AAD, but only from Kubernetes
+	approvedRemoveUserAssignedMSIIDs := make([]string, 0)
+	for _, item := range removeUserAssignedMSIIDs {
```
[the remark this answers is not in the document]
Comment (dmeytin) [reply pkg/mic/mic.go]: Done

File: pkg/mic/mic.go (hunk 4 above)
Comment (dmeytin) [reply pkg/mic/mic.go]: Done

File: docs/readmes/README.featureflags.md [hunk 10]
```diff
@@ -22,23 +22,8 @@ node/VMSS.
 Aad-pod-identity has a new flag clientQps which can be used to control the total number of client operations performed per second
 to the API server by MIC.
 
-## Block Instance Metadata flag
```
Comment (kkmsft) [inline docs/readmes/README.featureflags.md]: Looks like there is a need to rebase the code. The update to README from older patch got removed in this.

File: docs/readmes/README.featureflags.md [hunk 11]
```diff
@@ -22,23 +22,8 @@ node/VMSS.
...
-The `blockInstanceMetadata` flag for NMI will intercept any  requests to this
-endpoint from Pods which are not using host networking and return an HTTP 403
-Forbidden response. This flag is disabled by default to maximize compatibility.
-Users are encouraged to determine if this option is relevant and beneficial for
-their use cases.
+## ImmutableUserMSIs flag
+> Available from 1.5.4 release
+
+Aad-pod-identity has a new flag immutableUserMSIs which can be used to prevent deletetion of spcecified identitites from VM/VMSS.
+The list is comma separated.
```
Comment (kkmsft) [inline docs/readmes/README.featureflags.md]: Can you please show a simple example - /subscription/00000000-000... like that.. you can check the main README for example of the user identity pattern.

File: charts/aad-pod-identity/templates/mic-deployment.yaml [hunk 12]
```diff
@@ -62,6 +62,9 @@ spec:
           {{- if .Values.mic.clientQps }}
           - --clientQps={{ .Values.mic.clientQps }}
           {{- end }}
+          {{- if .Values.mic.immutableUserMSIs }}
+          - --immutableUserMSIs={{ .Values.mic.immutableUserMSIs }}
```
Comment (amirschw) [inline charts/aad-pod-identity/templates/mic-deployment.yaml]: consider passing the IDs as a list and concatenate them here with `join ","`

File: docs/readmes/README.featureflags.md (hunk 11 above)
Comment (dmeytin) [reply docs/readmes/README.featureflags.md]: The example is added: Example: 8307b873-f775-461a-9f16-45fa644b806a,9127b123-a116-654a-3f25-54af766a795b

File: docs/readmes/README.featureflags.md (hunk 10 above)
Comment (dmeytin) [reply docs/readmes/README.featureflags.md]: Fixed

File: charts/aad-pod-identity/templates/mic-deployment.yaml (hunk 12 above)
Comment (dmeytin) [reply charts/aad-pod-identity/templates/mic-deployment.yaml]: Done

File: charts/aad-pod-identity/README.md [hunk 13]
```diff
@@ -162,6 +162,7 @@ The following tables list the configurable parameters of the aad-pod-identity ch
...
 | `mic.probePort`                          | Override http liveliness probe port                                                                                                                                                                              | If not provided, default port is `8080`                  |
 | `mic.syncRetryDuration`                  | Override interval in seconds at which sync loop should periodically check for errors and reconcile                                                                                                               | If not provided, default value is `3600s`                |
+| `mic.immutableUserMSIs`                  | List of  user-defined identities that shouldn't be deleted from VM/VMSS. Note: the list should be **comma separated**                                                                                            | If not provided, default value is empty           |
```
Comment (amirschw) [inline charts/aad-pod-identity/README.md]: the `Note: ...` part can now be removed

File: charts/aad-pod-identity/values.yaml [hunk 14]
```diff
@@ -62,6 +62,12 @@ mic:
   # Override interval in seconds at which sync loop should periodically check for errors and reconcile (default is 3600s)
   syncRetryDuration: ""
 
+  # Override the defult value of immutable identities. Note: the list should be comma separated
```
Comment (amirschw) [inline charts/aad-pod-identity/values.yaml]: same here

File: charts/aad-pod-identity/README.md (hunk 13 above)
Comment (dmeytin) [reply charts/aad-pod-identity/README.md]: Correct!
Done

File: charts/aad-pod-identity/values.yaml (hunk 14 above)
Comment (dmeytin) [reply charts/aad-pod-identity/values.yaml]: Done

File: go.sum (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmeytin) [reply go.sum]: The changes been reverted

File: test/e2e/aadpodidentity_test.go [hunk 15]
```diff
@@ -133,6 +134,21 @@ var _ = Describe("Kubernetes cluster using aad-pod-identity", func() {
 		deleteAllIdentityValidator()
 	})
 
+	It("should not delete the Immutable Identity from vmss when the deployment is deleted", func() {
+		setUpIdentityAndDeployment(immutableIdentity, "", "1")
```
Comment (kkmsft) [inline test/e2e/aadpodidentity_test.go]: Ideally the e2e could:
1. Assign the immutable identity explicitly to the vmss/vm
2. Create an azure identity with the same immutable identity 
3. Run the identity validator check(ensure that a cycle has completed)
4. Now the identity validator is deleted (so it goes through a cycle of deletion) .
5. Check that count of  azureassignedidentity is 0 (confirming that the reconcile cycle is complete)
6. Check that the immutable identity is still present on the vm/vmss.

File: cmd/mic/main.go [hunk 16]
```diff
@@ -98,7 +103,10 @@ func main() {
 	config.Burst = int(clientQPS)
 	glog.Infof("Client QPS set to: %v. Burst to: %v", config.QPS, config.Burst)
 
-	micClient, err := mic.NewMICClient(cloudconfig, config, forceNamespaced, syncRetryDuration, &leaderElectionCfg, enableScaleFeatures, createDeleteBatch)
+	immutableUserMSIsList := strings.Split(immutableUserMSIs, ",")
+	glog.Infof("immutable identities are %v", immutableUserMSIsList)
```
Comment (kkmsft) [inline cmd/mic/main.go]: nit0: Can you please make this conditional, If the parameter is given, then only print. 
nit1: Can we have the first letter with caps.

**Pull request merged**